### PR TITLE
functorch.jvp support for autograd.Function

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90790
* #90789
* #90418
* #90415
* __->__ #90077

This PR adds functorch.jvp support for autograd.Function. It does so by
adding a jvp rule for custom_function_call.

For a regular PyTorch operation (like at::sin), the VariableType kernel:
- re-dispatches to at::sin
- calls the jvp rule for at::sin

The jvp rule for custom_function_call does just that. It constructs a
new autograd.Function (because the above logic already exists). Inside
the forward, it re-dispatches to custom_function_call. In the jvp rule,
it just calls whatever the jvp rule is supposed to be.

Since this logic is really close to the custom_function_call_grad, I
just put them together.

Test Plan:
- added jvp rules to the autograd.Function in autograd_function_db

### CODE REVIEW

**Comment by zou3519** [comment]
Not ready for review yet, getting some CI signal.

File: torch/_functorch/pyfunctorch.py [hunk 1]
```diff
@@ -122,12 +124,43 @@ def prev_grad_mode(self):
...
+        args, kwargs = self.lift(args, kwargs)
+        return kernel(self, *args, **kwargs)
+
+    # Jvp has custom lower because of the no_fwd_grad interaction
+    # See NOTE [grad and vjp interaction with no_grad] for related info.
+    # This logic is mirrored from C++ JvpInterpreterPtr::sendToNextInterpreter
+    def lower(self):
+        prev_fwd_grad_mode = self.prev_fwd_grad_mode()
+        if not self.prev_fwd_grad_mode:
+            return contextlib.nested(_set_fwd_grad_enabled(False), super().lower())
```
Comment (albanD) [inline torch/_functorch/pyfunctorch.py:149]: isn't contextlib.nested deprecated since python 2.7?

**Review by albanD — APPROVED** [review]
Sounds pretty good!

File: torch/_functorch/pyfunctorch.py (hunk 1 above)
Comment (zou3519) [reply torch/_functorch/pyfunctorch.py:149]: You're right, it's not in the Python 3 docs. I'll find an alternative

**Review by soulitzer — APPROVED** [review]
Cool!

**Pull request closed without merging**